server/src/service.rs

```diff
@@ -115,17 +115,61 @@ impl Service {
     #[zbus(out_args("unlocked", "prompt"))]
     pub async fn unlock(
         &mut self,
-        _objects: Vec<OwnedObjectPath>,
+        objects: Vec<OwnedObjectPath>,
     ) -> Result<(Vec<OwnedObjectPath>, ObjectPath), ServiceError> {
-        todo!()
+        let mut unlocked = Vec::new();
+        let prompt = ObjectPath::default();
+        let collections = self.collections.lock().await;
+
+        for object in &objects {
+            for collection in collections.iter() {
+                if object == collection.path() {
+                    if !collection.set_locked(false).await? {
+                        unlocked.push(object.clone());
+                    }
+                    tracing::info!("Collection: {} is locked.", object);
+                    break;
+                } else if let Some(item) = collection.item_from_path(object).await {
+                    if !item.set_locked(false).await? {
+                        unlocked.push(object.clone());
+                    }
+                    tracing::info!("Item: {} is locked.", object);
+                    break;
+                }
+                tracing::info!("Object: {} does not exist.", object);
+            }
+        }
+
+        Ok((unlocked, prompt))
     }
 
     #[zbus(out_args("locked", "prompt"))]
     pub async fn lock(
         &mut self,
-        _objects: Vec<OwnedObjectPath>,
+        objects: Vec<OwnedObjectPath>,
     ) -> Result<(Vec<OwnedObjectPath>, ObjectPath), ServiceError> {
-        todo!()
+        let mut locked = Vec::new();
+        let prompt = ObjectPath::default();
+        let collections = self.collections.lock().await;
+
+        for object in &objects {
+            for collection in collections.iter() {
+                if object == collection.path() {
+                    collection.set_locked(true).await?;
```

Comment:
Why is this not doing the same check as unlock? See above for fixing that

Reply:
> For unlock we need a way to identify when to issue a Prompt. But, for Lock we https://github.com/bilelmoussaoui/oo7/pull/150#discussion_r1826575597 need a prompt at all. This I'm sure :)

The spec disagree with you. 

![image](https://github.com/user-attachments/assets/a5f4d39c-8ba6-4339-bfd5-cc5b1888e1f1)

The current implementations don't use a prompt here, we might not need it if we agree with both the KDE/GNOME implementors, then it can easily be disabled. But for now, let us assume it is required to be fully compatible with the published specification.

But anyways, the lock/unlock DBus functions would call the "inner" one, so the `Unlock` could grow some functionality around it and the `Lock` one won't.

Reply:
Understood, thanks!

Reply:
Not sure why you are saying the implementation is wrong, it is incomplete because the whole Prompting is yet to be figured out with, once that is done both lock and unlock implementations as well as anything using Prompt. Hence me insisting on using Prompt:: default earlier

Reply:
Thanks for the reply.

Yeah it's incomplete. But,

> The purpose of service methods (Lock & Unlock) are to filter the objects locked/unlocked without prompt and send them out or send out a prompt without the objects.

And currently we're setting the locked status inside service methods. We shouldn't do this here. All we need to do this return objectspaths (locked/unlocked objects list or a prompt).

Since I can't get a feel of Lock with a prompt, I'm not sure when to issue a prompt and when to return locked objectpaths.

Reply:
> Who is supposed to modify the collection/item state then?

org.gnome.keyring.internal.Prompter.Callback [PromptReady](https://gitlab.gnome.org/GNOME/gcr/-/blob/master/gcr/org.gnome.keyring.Prompter.xml?ref_type=heads#L73)

Reply:
That will execute a task, the task will be created inside Lock and Unlock.